client_context.py

```diff
@@ -1,13 +1,104 @@
+from __future__ import annotations
+
+from typing import Optional
+import json
+
 from edge_service import EdgeService
 from featurehub_repository import FeatureHubRepository
+from fh_state_base_holder import FeatureStateBaseHolder
+
+
+class StrategyAttributeCountryName:
+    pass
+
+
+class StrategyAttributeDeviceName:
+    pass
+
+
+class StrategyAttributePlatformName:
+    pass
 
 
 class ClientContext:
     """holds client context"""
 
     def __init__(self, repo: FeatureHubRepository, edge: EdgeService):
         """"""
-        pass
+        self._repository = repo
+        self._attributes = {}
+
+    def user_key(self, value: str) -> ClientContext:
+        self._attributes['userkey'] = [value]
+        return self
+
+    def session_key(self, value: str) -> ClientContext:
+        self._attributes['session'] = [value]
+        return self
+
+    def country(self, value: StrategyAttributeCountryName) -> ClientContext:
+        self._attributes['country'] = [value]
+        return self
+
+    def device(self, value: StrategyAttributeDeviceName) -> ClientContext:
+        self._attributes['device'] = [value]
+        return self
+
+    def platform(self, value: StrategyAttributePlatformName) -> ClientContext:
+        self._attributes['platform'] = [value]
+        return self
+
+    def version(self, version: str) -> ClientContext:
+        self._attributes['version'] = [version]
+        return self
+
+    def attribute_value(self, key: str, value: str) -> ClientContext:
+        self._attributes[key] = [value]
+        return self
+
+    def attribute_values(self, key: str, values: list[str]) -> ClientContext:
+        self._attributes[key] = values
+        return self
+
+    def clear(self) -> ClientContext:
+        self._attributes.clear()
+        return self
+
+    def get_attr(self, key: str, default_value: Optional[str]) -> str:
+        if self._attributes[key]:
+            return self._attributes.get(key)[0]  # is this correct? what do we want to achieve?
+        return default_value
+
+    def default_percentage_key(self) -> str:
+        return self.get_attr('session') if self._attributes['session'] else self.get_attr('userkey')
+
+    def is_enabled(self, name: str) -> bool:
+        return self.feature(name).is_enabled()
+
+    def feature(self, name: str) -> FeatureStateBaseHolder:
+        return self._repository.feature(name)
+
+    def is_set(self, name: str) -> bool:
+        return self.feature(name).is_set()
+
+    def get_number(self, name: str) -> Optional[int, float]:
+        return self.feature(name).get_number()
```

Comment:
if your feature returns None, what is going to happen here? is there an Elvis Operator like ?.get_number() ?